### ISSUE TITLE
Provide serviceAccount token as token file in Slurm

### ISSUE THREAD

**Issue by antoinetran** [interTwin-eu/interlink-slurm-plugin#45]
# Short Description of the issue

When serviceAccount field is filled (even for default one called "default"), the pod in Kubernetes automatically receive the token to talk to kubernetes API
```
ls -al /var/run/secrets/kubernetes.io/serviceaccount/
drwxr-xr-x    2 root     root           100 Dec 19 13:21 ..2024_12_19_13_21_43.446791399
lrwxrwxrwx    1 root     root            31 Dec 19 13:21 ..data -> ..2024_12_19_13_21_43.446791399
lrwxrwxrwx    1 root     root            13 Dec 16 18:01 ca.crt -> ..data/ca.crt
lrwxrwxrwx    1 root     root            16 Dec 16 18:01 namespace -> ..data/namespace
lrwxrwxrwx    1 root     root            12 Dec 16 18:01 token -> ..data/token
```

Additionally, the kubelet automatically add environment variables that contains Kubernetes API address and port (thanks to https://github.com/kubernetes/kubernetes/blob/v1.9.0/pkg/kubelet/envvars/envvars.go#L45-L48):
```
KUBERNETES_SERVICE_PORT=443
KUBERNETES_SERVICE_HOST=10..... (some ip)
```
There are other KUBERNETES variables, but these two are the minimum for my test case, Argo, to be able to communicate to Kubernetes API.

## Environment

- Operating System: Debian 13 
- Other related components versions: Kind 1.25

## Steps to reproduce

deploy any pod, even with the default serviceAccount, this should not be in error
```
ls -al /var/run/secrets/kubernetes.io/serviceaccount/
env | grep KUBERNETES_
```

## Logs, stacktrace, or other symptoms

# Summary of proposed changes
See below

**Comment by antoinetran** [comment]
Ok I see a (long painful) path.

1. Add in InterLink current Role the ability to request serviceAccount token
```
- apiGroups: [""]
  resources: ["serviceaccounts/token"]
  verbs: ["create"]
```
2. In presence of this volume (this JSON is generated from volumes.String() and is a bit rearranged)
```
{
   "Name":"kube-api-access-kzzwz",
   "VolumeSource":"VolumeSource"{
      "Projected":"&ProjectedVolumeSource"{
         "VolumeProjection"{
            "VolumeProjection"{
               "ServiceAccountToken":"&ServiceAccountTokenProjection"{
                  "Audience":,
                  "ExpirationSeconds":*3607,
                  "Path":"token"
               },
            },
            "VolumeProjection"{
               "ConfigMap":"&ConfigMapProjection"{
                  "LocalObjectReference":"LocalObjectReference"{
                     "Name":"kube-root-ca.crt"
                  },
                  "Items":[
                     
                  ]"KeyToPath"{
                     "KeyToPath"{
                        "Key":"ca.crt",
                        "Path":"ca.crt",
                     }
                  },
               },
            },
            "VolumeProjection"{
               "DownwardAPI":"&DownwardAPIProjection"{
                  "DownwardAPIVolumeFile"{
                     "DownwardAPIVolumeFile"{
                        "Path":"namespace",
                        "FieldRef":"&ObjectFieldSelector"{
                           "APIVersion":v1,
                           "FieldPath":"metadata.namespace"
                        },
                     }
                  }
               },
            }
         },
         "DefaultMode":*420
      },
   }
}
```
- for `token`, in InterLink VK, request a service account token (see https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/1205-bound-service-account-tokens#example-flow) . For simplicity, `boundObjectRef` can be `null` and `validityDuration` should hard-coded to `99999h`, but it could be great if the first is attached to the pod UID. The second variable should stay at `99999h` because otherwise it would be a PITA to have InterLink VK to API to Slurm Plugin renew the token periodically.
This token file should be put in the pod creation request from InterLink VK to API, and API to (InterLink Slurm) Plugin.
The InterLink Slurm Plugin should put this content to pod directory `~/.interlink/pod-uid-blabla/volumes/kube-api-access-66fbd/token`
- for `ca.crt`, InterLink VK should request content of configMap kube-root-ca.crt present in its namespace (something like this below, but in HTTP request)
```
kubectl get configmap -n [namespace name] kube-root-ca.crt
NAME               DATA   AGE
kube-root-ca.crt   1      97m
```
The content should be in pod creation request from VK to API to Plugin.
The InterLink Slurm Plugin should put the content to a file in `~/.interlink/pod-uid-blabla/volumes/kube-api-access-66fbd/ca.crt`
- for `namespace`, in InterLink VK, if we have
`volume.Projected.Sources[0].DownwardAPI.Items[0].FieldRef.FieldPath == "metadata.namespace"` then we put the namespace where we are in pod request creation, from VK to API to Plugin.
The InterLink Slurm Plugin should put this content to pod directory `~/.interlink/pod-uid-blabla/volumes/kube-api-access-66fbd/namespace`
Note: this namespace is not interlink but it is the one where the serviceAccount exist.

3. Then the InterLink Slurm Plugin should add bind path to Singularity/Apptainer, from these file to the inside where the VolumeMounts want. Eg
```
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-66fbd
      readOnly: true
```
is translated to Singularity `--bind ~/.interlink/pod-uid-blabla/volumes/kube-api-access-66fbd:/var/run/secrets/kubernetes.io/serviceaccount`

**Comment by dciangot** [comment]
Just one quick comment. The projected accounts token will be rotated each hour.
I think there is no way around it, unless you pre create a token for a service account manually diabling the rotation (and the projection as well)

**Comment by antoinetran** [comment]
> Just one quick comment. The projected accounts token will be rotated each hour. I think there is no way around it, unless you pre create a token for a service account manually diabling the rotation (and the projection as well)

Hi Diego, the correct implementation would be for InterLink VK to do the same as what kubelet does: when 80% of expiration time is passed, then it should reask a new token, and provide them to InterLink API, then Plugin. The Slurm Plugin should have a way to trigger or undergo the rotation (either by listening to API with a port, or by polling regularly, or because it knows when a token should expire).
However that would be too much of work and need further discussion. The benefit of honoring the expiration time is, for the time being (InterLink still being experimental), less than just setting infinite time (something like 100 year). Which is what I am planning to do for now.

**Comment by antoinetran** [comment]
One thing I am not sure though is, after giving token to Slurm job, how to give it access to Kubernetes API? I suggest exposing Kubernetes API :
* to a NodePort Service. A random port 3XXXX is given
* to an Ingress Service. This is highly dependent on which Kubernetes cluster we are (AWS, GCP, baremetal, etc.)., what loadbalancing service it provides
* though a reverse SSH tunnel from VK to API, then the Slurm job would need access to API to a given port

I feel the easiest way and most compatible way to deploy is the first one (NodePort).

**Comment by dciangot** [comment]
> > Just one quick comment. The projected accounts token will be rotated each hour. I think there is no way around it, unless you pre create a token for a service account manually diabling the rotation (and the projection as well)
> 
> Hi Diego, the correct implementation would be for InterLink VK to do the same as what kubelet does: when 80% of expiration time is passed, then it should reask a new token, and provide them to InterLink API, then Plugin. The Slurm Plugin should have a way to trigger or undergo the rotation (either by listening to API with a port, or by polling regularly, or because it knows when a token should expire). However that would be too much of work and need further discussion. The benefit of honoring the expiration time is, for the time being (InterLink still being experimental), less than just setting infinite time (something like 100 year). Which is what I am planning to do for now.

As you say @antoinetran , it is so much work and so much of a burden to maintain that I'm not sure it is worth. Just to be aligned, did you check the path where you create the "infinite" service token manually so to avoid the projectedVolume one? I suspect anything else would be an hack around it.

Also, regarding argo workflows, I still think that a custom executor engine would be something to look at, I'll try to give it a look later this week.

**Comment by dciangot** [comment]
> One thing I am not sure though is, after giving token to Slurm job, how to give it access to Kubernetes API? I suggest exposing Kubernetes API :
> 
> * to a NodePort Service. A random port 3XXXX is given
> * to an Ingress Service. This is highly dependent on which Kubernetes cluster we are (AWS, GCP, baremetal, etc.)., what loadbalancing service it provides
> * though a reverse SSH tunnel from VK to API, then the Slurm job would need access to API to a given port
> 
> I feel the easiest way and most compatible way to deploy is the first one (NodePort).

Yes, either 1 or 2 I think are good options. That btw, are outside the plugin and interlink ballpark, right?

**Comment by antoinetran** [comment]
> > 
> 
> As you say @antoinetran , it is so much work and so much of a burden to maintain that I'm not sure it is worth. Just to be aligned, did you check the path where you create the "infinite" service token manually so to avoid the projectedVolume one? I suspect anything else would be an hack around it.

I don't think we can avoid projectedVolume. To me, the kubelet automatically converts serviceAccount to a projected volumes in the yaml. So to answer your question, the path field always exist.

**Comment by antoinetran** [comment]
> > 
> 
> Yes, either 1 or 2 I think are good options. That btw, are outside the plugin and interlink ballpark, right?

Yes, the more I think of it, the more I think any way to talk to Kubernetes API should be outside InterLink scope. Of course it can provide some example. I initially thought something could be done in the way InterLink already provides bastionSSH, so with a node port and ssh tunnel, we can access Kubernetes API. But that is not always the cleanest way (eg in Kind, the nodeport does not work unless we also map the host port).
